pandas/core/frame.py

```diff
@@ -8684,19 +8676,17 @@ def blk_func(values):
             with np.errstate(all="ignore"):
                 result = func(values)
 
-        if is_object_dtype(result.dtype):
+        if filter_type == "bool" and notna(result).all():
+            result = result.astype(np.bool_)
```

Comment:
can we just call coerce_to_dtypes?

Reply:
coerce_to_dtypes is about casting back to our `self.dtypes`, not to bool.

Reply:
hmm this is very messy (not your change), but what this is doing. better to overhaul no?

Reply:
This is specific to any/all, making sure we have bool-dtype.

There are _lots_ of things in _reduce that need an overhaul.  This one doesn't bother me.